airbyte-integrations/connectors/source-shopify/source_shopify/utils.py

```diff
@@ -323,3 +325,51 @@ def decorator(*args, **kwargs) -> Any:
             return func(*args, **kwargs)
 
         return decorator
+
+
+class LimitReducingErrorHandler(ErrorHandler):
+    """Custom error handler that reduces the request limit (of items/page) on 500 errors and retries.
+    limit is halved on each retry until it reaches 1. We still exponentially backoff on retries in addition to this.
+    """
+
+    def __init__(self, stream: "ShopifyStream", default_handler: ErrorHandler):
+        self.stream = stream
+        self.default_handler = default_handler
+
+    @property
+    def max_retries(self) -> Optional[int]:
+        return self.default_handler.max_retries
+
+    @property
+    def max_time(self) -> Optional[int]:
+        return self.default_handler.max_time
+
+    def interpret_response(self, response_or_exception: Optional[Union[requests.Response, Exception]]) -> ErrorResolution:
+        if isinstance(response_or_exception, requests.Response):
+            response = response_or_exception
+            if response.status_code == 500:
+                current_limit = self.stream._current_limit
+                if current_limit > 1:
+                    new_limit = max(1, current_limit // 2)
+                    self.stream._current_limit = new_limit
+                    new_url = self.update_limit_in_url(response.request.url, new_limit)
+                    response.request.url = new_url
```

Comment:
going off what I mentioned above, if we were to manipulate the `response.request.url` so that it has a smaller limit, then why do we need to make changes on `ShopifyStream` to update the limit if it's already been updating here to use a smaller limit and avoid the 500 error?